Update tutorial: Pseudobulk with decoupler and edgeR – Merge me for the GTA! 🚀 (#5947)
galaxyproject/training-material: topics/single-cell/tutorials/pseudobulk-analysis/tutorial.md

answer_histories:
  - label: "UseGalaxy.eu"
    history: https://usegalaxy.eu/u/dianitachj24/h/pseudo-bulk-edger-11-04-2025
    date: 2025-04-11
  - label: "UseGalaxy.eu"
    history: https://usegalaxy.eu/u/dianitachj24/h/pseudo-bulk-edger-pdcs-11-04-2025
    date: 2025-04-11
  - label: "UseGalaxy.eu"

>     history: https://usegalaxy.eu/u/dianitachj24/h/pseudo-bulk-edger-ncms-11-04-2025

Reviewer: why there are 3 answer histories?

Author: each answer history is for a different cluster within the dataset. So one is for pDCs, the other one NCM and the main one is for the complete AnnData file.

>
{: .agenda}

# Get the Data!

## Overview of the Data

> The dataset used in this tutorial is an `AnnData` object containing a subset of single-cell RNA-seq data, specifically including non-classical monocytes (NCMs) and plasmacytoid dendritic cells (pDCs) from bone marrow and peripheral blood samples. This subset was selected to illustrate differential gene expression between two immune cell types across two biologically distinct tissues.

Reviewer: Please cite the study/dataset

In this tutorial, we will focus on these two cell populations and explore how their expression profiles differ between bone marrow and blood using pseudobulk analysis.

Pseudobulk analysis is an advanced approach in single-cell data analysis. It involves aggregating single-cell expression data by group (e.g., by cell type and condition) to perform bulk-style analyses while preserving single-cell resolution through metadata.

> For this tutorial, we assume familiarity with standard single-cell data formats, such as `AnnData` or `Seurat` objects, and prior experience with single-cell workflows, particularly clustering and cell type annotation. We will use the `decoupler` tool to perform the pseudobulk aggregation and downstream differential analysis {% cite decoupler-pseudobulk %}.

Reviewer: Describe some basic numbers on total number cells or number of cells per condition etc. Otherwise, it is also good to show those numbers after data upload, at the very beginning of the analysis.

If you're new to these concepts, we recommend exploring the following tutorials before performing pseudobulk analysis:

> - [Clustering 3K PBMCs with Scanpy]({% link topics/single-cell/tutorials/scrna-scanpy-pbmc3k/tutorial.md %}): Learn how to perform clustering and cell type annotation using Galaxy's single-cell tools.

Reviewer: ```suggestion
- [Clustering 3K PBMCs with Scanpy]({% link topics/single-cell/tutorials/scrna-scanpy-pbmc3k/tutorial.md %}): Learn how to perform clustering and cell type annotation using Galaxy's Scanpy and Anndata tool suites.
```

In this step, our goal is to perform a "bioinformatic cell sorting" based on the annotated clusters of the single-cell data.

To start a pseudobulk analysis, ensure that the `AnnData` object contains all the necessary metadata for ***pseudobulking***. This includes key annotations such as `annotated` or `cell_type`, `condition`, `disease`, and `batch`. These labels may vary depending on how the dataset was originally annotated.

> 

Reviewer: Add Anndata inspect step (with general, obs and var inpsection) so that the user can see what exactly annotated in the anndata.

> Most importantly, the `AnnData` object must include a layer containing the raw gene expression counts. This layer is essential for accurate aggregation and downstream differential expression analysis. Raw counts are allows to generate accurate pseudobulk aggregates. Since single-cell data is typically normalized after annotation, it’s important to preserve the raw counts in the AnnData object before normalization steps if you would like to perform a pseudobulk analysis later on. These raw counts are directly used by tools like **Decoupler** to generate the pseudobulk count matrix. Note that normalized count matrices should not be used with Decoupler, even if the tool appears to process them successfully.

Reviewer: ```suggestion
Most importantly, the `AnnData` object must include a layer containing the raw gene expression counts. This layer is essential for accurate aggregation and downstream differential expression analysis. Raw counts are allows to generate accurate pseudobulk aggregates. Since single-cell data is typically normalized after annotation, it’s important to preserve the raw counts in the AnnData object before normalization steps if you would like to perform a pseudobulk analysis later on. These raw counts are directly used by tools like **Decoupler** to generate the pseudobulk count matrix. Note that normalized count matrices should not be used with **Decoupler**, even if the tool appears to process them successfully.
```

> >
> {: .solution}
>
{: .question}

# Subsetting Samples from the Original AnnData Object

In our initial analysis, we observed transcriptional differences between bone marrow and peripheral blood samples. However, since the dataset includes more than one immune cell type (pDCs and NCMs), it's not immediately clear which population is primarily driving these differences.

To gain further insight, it's helpful to perform pseudobulk analysis on each cell type separately. By subsetting the data by cluster and repeating the differential analysis (as we did with the full dataset at the start of this tutorial), we can better assess the individual contribution of each cell type.

In the following step, we’ll demonstrate how to subset the AnnData object by cell type before reapplying the pseudobulk workflow. In the hands-on example, we will filter for pDCs. To extract NCMs instead, you can follow the exact same steps, but set the *"Value"* parameter to `Non_Classical_Monocyte`, which corresponds to the annotation of that cell type.

> If you would like to extract **all annotated clusters at once**, for example to analyse each of them independently, refer to the tip box below titled **“Split AnnData object by cluster or other observation key into a collection”**.

Reviewer: ```suggestion
If you would like to extract all annotated clusters at once, for example to analyse each of them independently, refer to the tip box below titled **“Split AnnData object by cluster or other observation key into a collection”**.
```
